src/Sentry/Internal/MemoryMonitor.cs

```diff
@@ -0,0 +1,164 @@
+        {
+            return;
+        }
+
+        var usedMemory = Environment.WorkingSet;
+        if (!_dumpTrigger(usedMemory, _totalMemory))
+        {
+            return;
+        }
+
+        _options.HeapDumpDebouncer.RecordOccurence(eventTime);
+
+        var usedMemoryPercentage = ((double)usedMemory / _totalMemory) * 100;
+        _options.LogDebug("Total Memory: {0:N0} bytes", _totalMemory);
+        _options.LogDebug("Memory used: {0:N0} bytes ({1:N2}%)", usedMemory, usedMemoryPercentage);
+        _options.LogDebug("Automatic heap dump triggered");
+        _onCaptureDump();
+    }
+
+    internal void CaptureMemoryDump()
+    {
+        if (_options.DisableFileWrite)
+        {
+            _options.LogDebug("File write has been disabled via the options. Unable to create memory dump.");
+            return;
+        }
+
+        var dumpFile = TryGetDumpLocation();
+        if (dumpFile is null)
+        {
+            return;
+        }
+
+        var processId = Environment.ProcessId;
+        _options.LogInfo("Creating a memory dump for Process ID: {0}", processId);
+
+        // Check which patth to use for dotnet-gcdump. If it's been bundled with the application, it will be available
+        // in the `dotnet-gcdump` folder of the application directory. Otherwise we assume it has been installed globally.
+        var bundledToolPath = Path.Combine(AppContext.BaseDirectory, "dotnet-gcdump", "dotnet-gcdump.dll");
+        if (File.Exists(bundledToolPath))
+        {
+            _options.LogDebug($"Using bundled version of dotnet-gcdump from: {bundledToolPath}");
+        }
+        else
+        {
+            _options.LogDebug("Using global version of dotnet-gcdump");
+        }
+
+        var arguments = $"collect -p {processId} -o '{dumpFile}'";
+        var command = File.Exists(bundledToolPath)
+            ? $"dotnet {bundledToolPath} {arguments}"
+            : $"dotnet-gcdump {arguments}";
+        var startInfo = new ProcessStartInfo
+        {
+            FileName = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? "cmd.exe" : "/bin/bash",
+            Arguments = $"-c \"{command}\"",
+            RedirectStandardOutput = true,
+            UseShellExecute = false,
+            CreateNoWindow = true,
+        };
+
+        var process = new Process { StartInfo = startInfo };
+        process.Start();
+        while (!process.StandardOutput.EndOfStream)
+        {
+            if (process.StandardOutput.ReadLine() is { } line)
+            {
+                _options.LogDebug(line);
+            }
+        }
+
+        if (!_options.FileSystem.FileExists(dumpFile))
+        {
+            // if this happens, hopefully there would be more information in the standard output from the process above
+            _options.LogError("Unexpected error creating memory dump. Check debug logs for more information.");
+        }
+
+        _onDumpCollected(dumpFile);
```

Comment:
should we try/catch here and leave our internal method without error handling logic instead?

ideally we let things bubble out and capture at the top-most location (unless we want to recover and carry on)

Reply:
`_onDumpCollected(dumpFile)` is the callback once the dump has already successfully been created. It's mainly exposed to make it easier to unit test. In practice, this gets hooked up the `Hub.CaptureHeapDump` method which sends the heap dump to Sentry... and that does wrap it all in try/catch/log.

We also have `_onCaptureDump` which can be used to mock/intercept calls to `CaptureMemoryDump` in testing:
https://github.com/getsentry/sentry-dotnet/blob/2e6a0f131317fff04658f93aa0815d585b31f7d6/src/Sentry/Internal/MemoryMonitor.cs#L32

Ultimately that gets called via:
 `GarbageCollectionMonitor.MonitorGarbageCollection` -> `onGarbageCollected` (maps to `MemoryMonitor.CheckMemoryUsage`) -> `_onCaptureDump` (maps to `MemoryMonitor.CaptureMemoryDump`)

Any errors arising in that chain will ultimately bubble up to this now (set in the constructor for the MemoryMonitor):
https://github.com/getsentry/sentry-dotnet/blob/2dfd4f404ce15a3d16b993eef3ef1fee843b77de/src/Sentry/Internal/MemoryMonitor.cs#L36-L42